File: csm/ledger.py
"""Usage ledger: ~/.csm/ledger.json.

Every run appends a token record. The weekly budget guard works on a
rolling 7-day sum of *weighted* tokens:

    input + cache_creation + output + 0.1 * cache_read

This is a heuristic for subscription quota pressure, not billing (nothing
is billed - subscription auth only). Tune weekly_token_budget in config
after observing a normal week in `csm status`.
"""

from datetime import datetime, timedelta, timezone

from . import config


def load() -> list:
    return config.read_json(config.ledger_path(), [])


def weighted(usage: dict) -> int:
    return round(
        usage.get("input_tokens", 0)
        + usage.get("cache_creation_input_tokens", 0)
        + usage.get("output_tokens", 0)
        + 0.1 * usage.get("cache_read_input_tokens", 0)
    )


def append(item_id: str, project: str, model: str, usage: dict) -> None:
    records = load()
    records.append({
        "ts": datetime.now(timezone.utc).isoformat(timespec="seconds"),
        "item_id": item_id,
        "project": project,
        "model": model,
        "usage": usage,
        "weighted": weighted(usage),
    })
    config.write_json(config.ledger_path(), records)


def weekly_spend(records: list | None = None) -> int:
    if records is None:
        records = load()
    cutoff = datetime.now(timezone.utc) - timedelta(days=7)
    total = 0
    for rec in records:
        try:
            ts = datetime.fromisoformat(rec["ts"])
        except (KeyError, ValueError):
            continue
        if ts >= cutoff:
            total += rec.get("weighted", 0)
    return total

File: internal/ledger/ledger.go
// Package ledger records token usage per run. The weekly budget guard works
// on a rolling 7-day sum of *weighted* tokens:
//
//	input + cache_creation + output + 0.1 * cache_read
//
// This is a heuristic for subscription quota pressure, not billing (nothing
// is billed - subscription auth only).
package ledger

import (
	"math"
	"time"

	"github.com/EkinBarisC/claude-session-manager/internal/config"
)

type Record struct {
	TS       string         `json:"ts"`
	ItemID   string         `json:"item_id"`
	Project  string         `json:"project"`
	Model    string         `json:"model"`
	Usage    map[string]int `json:"usage"`
	Weighted int            `json:"weighted"`
}

func Load() []Record {
	var records []Record
	config.ReadJSON(config.LedgerPath(), &records)
	return records
}

func Weighted(usage map[string]int) int {
	return int(math.Round(
		float64(usage["input_tokens"]) +
			float64(usage["cache_creation_input_tokens"]) +
			float64(usage["output_tokens"]) +
			0.1*float64(usage["cache_read_input_tokens"])))
}

func Append(itemID, project, model string, usage map[string]int) error {
	records := append(Load(), Record{
		TS:       time.Now().UTC().Format("2006-01-02T15:04:05+00:00"),
		ItemID:   itemID,
		Project:  project,
		Model:    model,
		Usage:    usage,
		Weighted: Weighted(usage),
	})
	return config.WriteJSON(config.LedgerPath(), records)
}

// WeeklySpend sums weighted tokens over the last 7 days.
func WeeklySpend() int {
	cutoff := time.Now().UTC().AddDate(0, 0, -7)
	total := 0
	for _, rec := range Load() {
		ts, err := time.Parse(time.RFC3339, rec.TS)
		if err != nil {
			continue
		}
		if !ts.Before(cutoff) {
			total += rec.Weighted
		}
	}
	return total
}
